### ISSUE TITLE
Is there a faster implementation of Twofish available?

### ISSUE THREAD

**Issue by Logan007** [ntop/n2n#247]
n2n's default cipher Twofish seems to be the slowest in all comparisons using `tools/n2n-benchmark`. Older [comparisons](https://panthema.net/2008/0714-cryptography-speedtest-comparison/) suggest that Twofish does not necessarily has to be slower than software AES.

Would anybody volunteer to check for faster implementations? I found [this](https://github.com/drewcsillag/twofish) older one which is [referenced](https://www.schneier.com/academic/twofish/download.html) (fourth line from the bottom) by Twofish's author Bruce Schneier. In a really quick attempt, it did not let me compile – as opposed to [this fork](https://github.com/fudanchii/twofish) of it.

I guess it could be moved to `twofish.c` if comparative testing should speedwise prove it superior.

**Comment by lucktu** [comment]
By default, twofish is indeed slower than AES in most cases.
If you want to do some tests, please enter the QQ group: 5804301

**Comment by Logan007** [comment]
@lucktu Thank you for offering testing. The currently used Twofish implementation shows roughly 55 percent of software AES' (openSSL) performance only¹ – on Intel as well as ARM. What platforms do you have available? MIPS, SPARC and maybe RISC-V would be interesting. I would be curious how current Twofish compares to software AES on those.

However, somebody would need to port the above mentioned fork (or an even faster implementation – let us know) to `twofish.c` maintaining full compatibility for – as a next step – conducting a comparison between these different Twofishes under the same conditions on different platforms.
___________________
¹ This might not be a too fair comparison as openSSL's software implementations (not the hardwardware accelerated) surely employ highly platform-optimized  implementations (possibly MMX, SSE, AVX, NEON) – a road n2n itself probably does not go down. Anyway, if there is a faster standard C Twofish implementation available, we should use it.

**Comment by lucktu** [comment]
I'm a little busy at the moment. I don't have much time for tests. 

But I've done a lot of testing before, both here and on my blog (mainly for mips) .

**Comment by Logan007** [comment]
@lucktu Good to hear that you have MIPS available for testing... we might need it later whenever another implementation of Twofish shows up.

From [this example](https://github.com/ntop/n2n/issues/230#issuecomment-619683650) I get that Twofish compared to AES seems to perform even worse on MIPS (only 45 percent of AES' performance). Another reason to look out for a faster implementation.

Let me add, as a postscript, that your n2n-related articles on your webpage are quite an impressive resource!

**Comment by Logan007** [comment]
I got a little deeper into [Twofish](https://www.schneier.com/academic/paperfiles/paper-twofish-paper.pdf) last night and found that the _g_ function which is crucial to speed, already is extremely optimized in n2n's implementation as well as in Drew Scillag's version mentioned above. 

I updated my [fork](https://github.com/Logan007/twofish) of the version that is based on Drew Scillag's implementation. Now, compilation should be easier and the parameters for the `twofish-benchmark` now are comparable to n2n's `tools/n2n-benchmark`. It turns out that n2n's implementation (85 MB/s) indeed seems somewhat slower than Drew Scillag's (140 MB/s) on that same computer. If my measurements are correctly… Also, we should not forget about the additional cost of n2n's various activities around the pure cipher in the `transform_tf.c` file which are not reflected in the other `twofish-benchmark`.

So, I will give it a shot and report back – not too soon as this will require some time and is not high on my priority list.

_They both already take advantage of the 4K look-up table which is mentioned in chapter 5. I wonder if a 512K look-up table would help (two 16x32 bit tables) requiring two look-ups and one exclusive-or only… I just fear that memory access is too costly as table size grows beyond a good "cachability"._

_Another thought: Maybe, Twofish's CBC **de**cryption could benefit from a SSE implementation that would handle four blocks at a time. The look-up in memory probably remains the bottleneck. Still thinking…_
